File: 785-is-graph-bipartite/785-is-graph-bipartite.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        /*
        we are getting the adj list.
        
        which will be of the form, 0 -> {1, 2, 3}, 1 -> {0, 2}, 2 -> {0. 1, 3}
        and 3 -> {0, 2}.
        */
        int n = graph.size();
        vector<int> color(n, -1);
        
        for(int i=0;i<n;i++)
            if(color[i] == -1)
                if(!bfsCheck(graph, i, color)) return false;
        
        return true;
    }
    
    bool bfsCheck(vector<vector<int>>& graph, int node, vector<int>& color) {
        queue<int> q;
        q.push(node);
        
        color[node] = 1;
        
        while(!q.empty()) {
            int curr = q.front();
            q.pop();
            
            for(int u : graph[curr]) {
                if(color[u] == -1) {
                    color[u] = 1 - color[curr];
                    q.push(u);
                }
                else if(color[u] == color[curr]) return false;
            }
        }
        return true;
    }
};

File: 785-is-graph-bipartite/785-is-graph-bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        //DFS
        int n = graph.length;
        int[] color = new int[n];
        
        for(int i=0;i<n;i++) color[i] = -1;
        
        for(int i=0;i<n;i++)
            if(color[i] == -1)
                if(dfsCheck(i, graph, color) == false) return false;
            
        return true;
    }
    
    boolean dfsCheck(int node, int[][] graph, int[] color) {
        if(color[node] == -1) color[node] = 0;
        
        for(int u : graph[node]) {
            if(color[u] == -1) {
                color[u] = 1 - color[node];
                if(dfsCheck(u, graph, color) == false) return false;
            }
            else {
                if(color[u] == color[node]) return false;
            }
        }
        return true;
    }
}